Fix division by zero in ColdBlockInfo::inferFromEdgeProfile

karim-alweheshy opened the pull request:
## Summary

This PR fixes a division by zero crash in the Swift compiler's profile-guided optimization implementation and provides user-configurable strategies for handling zero-count blocks, inspired by similar improvements in LLVM.

## Background

LLVM recently addressed similar issues with zero-count blocks in profile-guided optimization (see [LLVM PR #154437](https://github.com/llvm/llvm-project/pull/154437)), which added configurable strategies for handling blocks that were instrumented but never executed during profiling. This change brings similar capabilities to Swift.

## Problem

The issue manifests in `ColdBlockInfo::inferFromEdgeProfile` when calculating taken probabilities:
```cpp
double takenProbability = succCount[i].getValue() / (double)totalCount.getValue();
```

When `totalCount` is 0 (block was instrumented but never executed), this causes a division by zero crash during SIL optimization passes.

## Solution

Added a check for `totalCount.getValue() < 1` before the division operation, with user-configurable strategies for handling zero-count blocks:

### Command-Line Option
```bash
-sil-zero-count-strategy=<strategy>
```

### Available Strategies

1. **`conservative` (default)**: Skip inference, let other heuristics decide
   - Safe default behavior that prevents crashes
   - Makes no assumptions about missing execution data
   - Follows the same pattern as other Swift compiler components

2. **`optimistic`**: Assume zero-count blocks are cold
   - Logical assumption: zero execution counts = never executed = cold code
   - Useful for aggressive size optimization when profile data is trusted
   - Aligns with the reasoning that warm/important code would have execution counts

## Changes

1. **ColdBlockInfo.cpp**: 
   - Added zero-count check with configurable strategies
   - Command-line option with enum-based strategy selection
   - Debug logging for strategy decisions

2. **cold_block_zero_count.sil**: 
   - Comprehensive test cases covering both strategies
   - Tests blocks with zero execution counts on all branches  
   - Tests mixed zero/non-zero execution counts
   - Tests inlining scenarios with zero-count blocks

## Design Rationale

The two-strategy approach is based on logical reasoning about zero-count scenarios:
- **Zero counts indicate**: either "truly not executed" or "missing profile data"
- **Conservative approach**: When uncertain, make no assumptions (safe default)
- **Optimistic approach**: Assume zero counts mean cold code (aggressive optimization)

No "aggressive/warm" strategy is included because warm/important code would logically have execution counts during profiling.

## Testing

The test reproduces the division by zero condition and verifies both strategies handle it gracefully:
```bash
# Conservative strategy (default)
%target-sil-opt %s -performance-inline

# Optimistic strategy  
%target-sil-opt %s -performance-inline -sil-zero-count-strategy=optimistic
```

Profile data with zero counts is common for error handling paths, platform-specific code, and rarely executed functions. This fix ensures the compiler handles such cases without crashing while providing optimization flexibility.

## Compatibility

- Backward compatible: conservative strategy is the default
- Follows existing Swift patterns for command-line options
- Aligns with LLVM's approach to profile-guided optimization control

kavon reviewed (changes requested):
Thanks! Overall this looks great. Do you think you could squash this into one commit?

kavon commented on lib/SILOptimizer/Analysis/ColdBlockInfo.cpp:
```diff
@@ -221,31 +238,99 @@ bool ColdBlockInfo::inferFromEdgeProfile(SILBasicBlock *BB) {
+  // future propagation algorithms, which need to know where data is missing.
+  bool hasAnyNonZeroCount = false;
+  bool hasAnyMissingData = false;
+  SmallVector<std::optional<ProfileCounter>, 2> counters;
```
`std::optional<ProfileCounter>` is redundant here. ProfileCounter already has a representation of "missing"

kavon commented on lib/SILOptimizer/Analysis/ColdBlockInfo.cpp:
```diff
@@ -221,31 +238,99 @@ bool ColdBlockInfo::inferFromEdgeProfile(SILBasicBlock *BB) {
+        << "ColdBlockInfo: treating missing profile data as zero for "
+        << toString(BB->getSuccessors()[i])
+        << " (optimistic strategy - found non-zero counts on other edges)\n");
+    }
```
This is applying the optimistic strategy unconditionally. I would expect it to be aligned with the flag.

kavon commented on lib/SILOptimizer/Analysis/ColdBlockInfo.cpp:
```diff
@@ -221,31 +238,99 @@ bool ColdBlockInfo::inferFromEdgeProfile(SILBasicBlock *BB) {
-  // especially since we only have two temperatures.
+  // especially since we only have two temperatures (cold/warm).
+  // TODO: With propagation algorithm, this limitation can be removed to support
+  // multi-way branches (switch statements) and arbitrary successor counts.
```
nit: inaccurate addition to this comment

There already is propagation. The reason for this limitation is hinted at in the original comment: there's only two temperatures being set and propagated for blocks. We set a block cold only if it's taken with very low probability, currently under 3%, otherwise it's just "warm".  If a block has 100 successors and they are all taken with equal probability of 1%, then we'd mistakenly mark all blocks cold.

Definitely could use a better strategy here, but I think for this PR let's stay focused on fixing the divide-by-zero and having the "assume cold for missing data".

kavon reviewed (changes requested):
Please ping me when you think this is ready for another round of review! Thanks.

karim-alweheshy commented:
@kavon ready for review

jonmorganreddit commented:
Hi @kavon and @eeckstein! Is there any chance we can get this fix merged? We routinely hit this bug and would love to have a fix in. Thanks!

eeckstein commented:
> Is there any chance we can get this fix merged? We routinely hit this bug and would love to have a fix in. Thanks!

@kavon?

keith commented:
@swift-ci please test

karim-alweheshy commented:
@swift-ci please test

xavgru12 commented:
@swift-ci please test

karim-alweheshy commented:
@swift-ci please test

karim-alweheshy committed:
Fix division by zero in ColdBlockInfo::inferFromEdgeProfile

karim-alweheshy commented:
@kavon I addressed the outstanding squash request and rebased this onto current `main`. The branch is now one commit (`0ed82406a224`), all inline threads remain resolved, and GitHub reports it mergeable. The parser conflict was adapted to the current `cond_br` API (no branch arguments), retaining `!true_count` / `!false_count` parsing and passing the counters to the current `createCondBranch` overload. `git diff --check` and touched-line clang-format are clean. As a negative control, Apple Swift 6.3.2 still rejects those counters in the new SIL test, confirming that the parser coverage is necessary. This exact rebased head now needs an upstream compiler build and test run; could you please re-review and trigger Swift CI?
